Transformations: Build help content for all standard transformers

baldm0mma opened the pull request:
**What is this feature?**

This feature adds helpful content to the transformation editors when the "help" button is clicked. Previous to this, it simply rendered the message: "Go the transformation documentation for more."

**Why do we need this feature?**

In-app help to understand transformations is a core user experience.

**Who is this feature for?**

Anyone using transformations.

**Which issue(s) does this PR fix?**:

Fixes #74062

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [na] If this is a pre-GA feature, it is behind a feature toggle.
- [na] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

https://github.com/grafana/grafana/assets/46619047/471b2aad-53cc-42eb-8b4f-8ccdf19ff09c

baldm0mma committed:
baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ add getDisplayContent

baldm0mma committed:
baldm0mma/transformation_info_modal/ reamove OperationRowHelp

baldm0mma committed:
baldm0mma/transformation_info_modal/ update prepMarkdown

baldm0mma committed:
baldm0mma/transformation_help_modal/ add example mkdwn

baldm0mma committed:
baldm0mma/transformation_help_modal/ add examples anno

baldm0mma commented:
This tackles #74062

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ add calFieldHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ remove file

baldm0mma committed:
baldm0mma/transformation_info_modal/ add CalculateFieldHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add TransformationEditorHelperModal

baldm0mma committed:
baldm0mma/transformation_info_modal/ add getHelperContent

baldm0mma committed:
baldm0mma/transformation_info_modal/ move TransformationEditorHelperModal

baldm0mma committed:
baldm0mma/transformation_info_modal/ add ReactElement

baldm0mma committed:
baldm0mma/transformations_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ update props for TransformationEditorHelperModal

baldm0mma committed:
baldm0mma/transformation_info_modal/ test

baldm0mma committed:
test

baldm0mma committed:
baldm0mma/transformation_info_modal/ reset vscode auth

baldm0mma committed:
baldm0mma/transformation_info_modal/ update getHelperContent

baldm0mma committed:
baldm0mma/transformations_info_modal/ add items for rebuilding markdown to react component

baldm0mma committed:
test

baldm0mma committed:
test

baldm0mma committed:
baldm0mma/transformations_info_modal/ remove comment in TransformationOperationRow

baldm0mma committed:
test

baldm0mma committed:
baldm0mma/transformation_info_modal/ add element to typing

baldm0mma committed:
baldm0mma/transformation_info_modal/ update getHelperContent with help component

baldm0mma committed:
baldm0mma/transformation_info_modal/ update modal title

baldm0mma committed:
baldm0mma/transformation_info_modal/ move logic to tranforms

baldm0mma committed:
baldm0mma/transformation_info_modal/ remove unused comments

baldm0mma committed:
baldm0mma/transformation_info_modal/ update standard reg

baldm0mma committed:
baldm0mma/trfansformation_info_modal/ remove context styling

baldm0mma committed:
baldm0mma/transformation_info_modal/ add content for CalculateFieldHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add ConcatenateHelper

baldm0mma committed:
baldm0mma/transformations_info_modal/ add ConcatenateHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add ConfigFromQueryHelper

baldm0mma committed:
baldm0mma/transformations/info_modal/ add ConvertFieldTypeHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add CreateHeatmapHelp

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ add ExtractFieldsHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add FilterFieldsByNameHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add FilterFieldsByNameHelper

baldm0mma committed:
Merge branch 'baldm0mma/transformation_info_modal' of github.com:grafana/grafana into baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ add OperationRowHelp to parse markdown

baldm0mma committed:
baldm0mma/transformation_info_modal/ remove commented out markdown

baldm0mma committed:
baldm0mma/transformations_info_modal/ remove comment

baldm0mma committed:
baldm0mma/transformation_info_modal/ update configFromQueryTransformRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update ConcatenateHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ update ConvertFieldTypeHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ update ExtractFieldsHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ update FieldLookupHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ update TransformationEditorHelperModal with determineContentDisplay()

baldm0mma committed:
baldm0mma/transformation_info_modal/ add FilterByRefIdHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add FilterByValueHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ update file ext

baldm0mma committed:
baldm0mma/transformation_info_modal/ add GroupByHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add HistogramHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add JoinByFieldHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add JoinByLabelsHelper

baldm0mma committed:
baldm0mma/transformations_info_modal/ add LabelsToFieldsHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add LimitHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add MergeHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add PartitionByValuesHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add ReduceHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add RenameByRegexHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add SeriesToRowsHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ add SortByHelper

baldm0mma committed:
baldm0mma/transformations_info_modal/ add spatial

baldm0mma committed:
baldm0mma/transformation_info_modal/ update to markdown

baldm0mma committed:
baldm0mma/transformation_info_modal/ remove comments

baldm0mma committed:
baldm0mma/transformation_info_modal/ update boarder with theme

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ add test suite

baldm0mma committed:
baldm0mma/transformation_info_modal/ update tests

baldm0mma committed:
baldm0mma/transformation_info_modal/ update test to run on all transformations

baldm0mma committed:
baldm0mma/transformation_info_modaol/ add TimeSeriesTableHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ update concatenateHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ update calculateFieldHelper

baldm0mma committed:
baldmomma/transformation_info_modal/ update naming

baldm0mma committed:
baldm0mma/transformation_info_modal/ update getLinkToDocs location

baldm0mma committed:
baldm0mma/transformation_info_modal/ update docs

baldm0mma committed:
baldm0mma/transformation_info_modal/ update language

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ rename test

baldm0mma committed:
baldm0mma/transformation_info_modal/ update naming 2

codeincarnate commented on public/app/features/transformers/helpers/index.ts:
```diff
@@ -0,0 +1,29 @@
+export { calculateFieldHelper } from './calculateFieldHelper';
```
A note too that regardless we should load these dynamically so we're not adding to bundled code size.

baldm0mma replied:
Great points. 1) Much of this is already in the docs markdown, as you said, but I don't believe there's any way to target specific parts of that markdown with javascript to then use in specific places. I believe there are markdown extensions that can add the ability to include/add markdown fragment files into larger files, which would mean we could break up the files and use them separately, but not without adding another dependency. Plus, I would assume whatever parser looks at these for the website content would be real mad about it. But yes, having this in a single place is certainly best. I'll think about this more. 2) These are all functions that return pretty slim multi-line strings, which doesn't seem like very large overhead. Is it worth loading dynamically? Also, I though through it a bit, but couldn't come up with a good way to do that, so I'd love any suggestions. :)

baldm0mma committed:
baldm0mma/transformation_info_modal/ update to helper

baldm0mma committed:
baldm0mma/transformation_info_modal/ update docs

ryantxu commented:
I agree with @codeincarnate -- this looks good, but will almost certainly be out-of-sync instantly.  Can we explore any method that would help with that?  I suspect the best approach involves codegen :)

imatwawana commented on public/app/features/transformers/helpers/labelsToFieldsHelper.ts:
```diff
@@ -0,0 +1,63 @@
+export const labelsToFieldsHelper = () => {
+  return `
+  Use this transformation to change time series results that include labels or tags into a table where each label keys and values are included in the table result. The labels can be displayed either as columns or as row values.
```
```suggestion
  Use this transformation to change time series results that include labels or tags into a table where each label's keys and values are included in the table result. The labels can be displayed as either columns or row values.
```

imatwawana commented on public/app/features/transformers/helpers/partitionByValuesHelper.ts:
```diff
@@ -0,0 +1,33 @@
+export const partitionByValuesHelper = () => {
+  return `
+  Use this transformation to eliminate the need for multiple queries to the same datasource with different 'WHERE' clauses when graphing multiple series. Consider a metrics SQL table with the following data:
```
```suggestion
  Use this transformation to eliminate the need for multiple queries to the same data source with different 'WHERE' clauses when graphing multiple series. Consider a metrics SQL table with the following data:
```

imatwawana commented on public/app/features/transformers/helpers/rowsToFieldsHelper.ts:
```diff
@@ -0,0 +1,62 @@
+export const rowsToFieldsHelper = () => {
+  return `
+  Use this transformation to convert rows into separate fields. This can be useful as fields can be styled and configured individually. It can also use additional fields as sources for dynamic field configuration or map them to field labels. The additional labels can then be used to define better display names for the resulting fields.
```
```suggestion
  Use this transformation to convert rows into separate fields. This can be useful because fields can be styled and configured individually. It can also use additional fields as sources for dynamic field configuration or map them to field labels. The additional labels can then be used to define better display names for the resulting fields.
```

imatwawana commented on public/app/features/transformers/helpers/sortByHelper.ts:
```diff
@@ -0,0 +1,5 @@
+export const sortByHelper = () => {
+  return `
+  Use this transformation to sort each frame by the configured field. When 'reverse' is checked, the values will return in the opposite order.
```
```suggestion
  Use this transformation to sort each frame by the configured field. When the **Reverse** switch is on, the values will return in the opposite order.
```

imatwawana commented on public/app/features/transformers/helpers/timeSeriesTableHelper.ts:
```diff
@@ -0,0 +1,7 @@
+export const timeSeriesTableHelper = () => {
+  return `
+  Use this transformation to convert time series result into a table, converting time series data frame into a “Trend” field. “Trend” field can then be rendered using [sparkline cell type](https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/table/#sparkline), producing an inline sparkline for each table row. If there are multiple time series queries, each will result in a ...
```
```suggestion
  Use this transformation to convert time series results into a table, converting a time series data frame into a trend visualization field. A trend field can then be rendered using the [sparkline cell type](https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/table/#sparkline), producing an inline sparkline for each table row. If there are multiple time series queries, each will result in a separate table data frame. These can be joined using join or merge transforms to produce a single table with multiple sparklines per row.
```

imatwawana reviewed (commented):
I made some wording edits and tried to replicate whatever I've updated in the docs file in the helper files. I left a question on of them where the wording in the docs file is different from the helper file and I'm not 100% sure which meaning you're aiming for.

baldm0mma committed:
Apply suggestions from @imatwawana

Co-authored-by: Isabel <76437239+imatwawana@users.noreply.github.com>

baldm0mma committed:
baldm0mma/transformation_info_modal/ update convertFieldTypeHelper capitalization

jdbaldry committed:
Hack around with generating Markdown from transformations helpers (#75677)

```console
$ cd docs
$ make sources/panels-visualizations/query-transform-data/transform-data/index.md
$ make docs
```

Browse to http://localhost:3003/docs/grafana/latest/panels-visualizations/query-transform-data/transform-data/

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

baldm0mma committed:
baldm0mma/transformation_info_modal/ add calculateFieldHelper

baldm0mma committed:
baldm0mma/transformation_info_modal/ update concatenate

baldm0mma committed:
baldm0mma/transformation_info_modal/ add fieldLookup

baldm0mma committed:
baldm0mma/transformation_info_modal/ add filterFieldsByName

baldm0mma committed:
baldm0mma/transformation_info_modal/ add joinByField

baldm0mma committed:
baldm0mma/transformation_info_modal/ update renameByRegex

baldm0mma committed:
baldm0mma/transformation_info_modal/ update getHelperContent

baldm0mma committed:
baldm0mma/transformation_info_modal/ update calculateFieldTransformRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update concatenateTransformRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update configFromDataTransformer

baldm0mma committed:
baldm0mma/transformation_info_modal/ update extractFieldsTransformRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update formatTimeTransformerRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update heatmapTransformRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update mergeTransformerRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update timeSeriesTableTransformRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update doc builder

baldm0mma committed:
baldm0mma/transformation_info_modal/ update docs

baldm0mma committed:
baldm0mma/transformation_info_modal/ remove console log

baldm0mma committed:
baldm0mma/transformation_info_modal/ update console log

baldm0mma committed:
baldm0mma/transformation_info_modal/ update gitignore

baldm0mma committed:
baldm0mma/transformation_info_modal/ update calculateField

baldm0mma committed:
baldm0mma/transformation_info_modal/ add links

baldm0mma committed:
baldm0mma/transformation_info_modal/ update template

baldm0mma committed:
baldm0mma/transformation_info_modal/ update template with images

baldm0mma committed:
baldm0mma/transformation_info_modal/ build markdown

baldm0mma committed:
baldm0mma/transformations_info_modal/update with links

baldm0mma committed:
baldm0mma/transformation_info_modal/ update links

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ apdate FormatTimeTransfomerEditor

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update HeatmapTransformerEditor

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update ConfigFromQueryTransformerEditor

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update CalculateFieldTransformerEditor

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update ConcatenateTransformerEditor

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update FilterByNameTransformerEditor

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update FilterByRefIdTransformerEditor

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update GroupByTransformerEditor

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update GroupingToMatrixTransformerEditor get helper

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update SeriesToFieldsTransformerEditor get helper

baldm0mma committed:
baldm0mma/transformation_indo_modal/ update LabelsAsFieldsTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update MergeTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update OrganizeFieldsTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update ReduceTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update RenameByRegexTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update SeriesToRowsTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update SortByTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update extractFieldsTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update FilterByValueTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update JoinByLabelsTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update FieldLookupTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update PartitionByValuesEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update PrepareTimeSeriesEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update RowsToFieldsTransformerEditor

baldm0mma committed:
baldm0mma/transformation_info_modal/ update spatialTransformRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update timeSeriesTableTransformRegistryItem

baldm0mma committed:
baldm0mma/transformation_info_modal/ update to helperDocs

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ update built markdown

baldm0mma committed:
baldm0mma/transformation_info_modal/ run prettier

baldm0mma committed:
baldm0mma/transformation_info_modal/ update comments in getTransformationContent

baldm0mma committed:
baldm0mma/transformation_info_modal/ add removeMarkdownAnchorLinks

baldm0mma committed:
baldm0mma/transformation_info_modal/ update annos in getTransformationContent

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ add requisite url prefix to markdown images

baldm0mma committed:
baldm0mma/transformation_info_modal/ add cleanMarkdownOfUnwantedSyntax

baldm0mma committed:
baldm0mma/transformation_info_modal/ update annos in content.ts

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

baldm0mma committed:
baldm0mma/transformation_info_modal/ update e2e tests with always-visiable trabnsformation

baldm0mma committed:
baldm0mma/transformation_info_modal/ remove note

baldm0mma commented on e2e/panels-suite/panelEdit_base.spec.ts:
```diff
@@ -39,7 +39,7 @@ describe('Panel edit tests', () => {
           expect(li.text()).equals('Transform data0'); // there's no transform so therefore Transform + 0
         });
-        e2e.components.Transforms.card('Merge').scrollIntoView().should('be.visible');
+        e2e.components.Transforms.card('Add field from calculation').scrollIntoView().should('be.visible');
```
The `merge` transformation is no longer "visible" since there aren't 2 series available for `merge` to be "pickable."

baldm0mma replied:
That's what I thought as well, but it is nevertheless not found the `be.visable` assertion. I think the `Add field from calculation` is a better de facto test, since it will always be available, but I can look into building out a better test for merge - or series of tests for the other transformations - if we think it's worth it. Thoughts?

baldm0mma replied:
Looks like it was a naming issue after all. I updated the test to reflect the new title for merge: `Merge` -> `Merge series/tables`.

baldm0mma commented on public/app/features/transformers/docs/content.ts:
```diff
@@ -63,7 +63,7 @@ export const transformationDocsContent: TransformationDocsContentType = {
   ${buildImageContent(
     '/static/img/docs/transformations/add-field-from-calc-stat-example-7-0.png',
     imageRenderType,
-    this.name
```
`this` is not bound in the correct context when the `buildImageContent` function is being called, so name is `undefined`.

baldm0mma committed:
baldm0mma/transformation_info_modal/ add annotation in getTransformationContent

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

oscarkilhed commented on public/app/features/transformers/spatial/SpatialTransformerEditor.tsx:
```diff
@@ -165,4 +167,5 @@ export const spatialTransformRegistryItem: TransformerRegistryItem<SpatialTransf
   description: spatialTransformer.description,
   state: PluginState.alpha,
   categories: new Set([TransformerCategory.PerformSpatialOperations]),
+  help: getTransformationContent(DataTransformerID.spatial).helperDocs,
```
I noticed the spatial docs were quite sparse. Probably not something for this PR, but we should make a follow up PR for it.

baldm0mma replied:
Agreed. That's the final piece of this epic: expanding/correcting the docs themselves.

baldm0mma committed:
baldm0mma/transformation_info_modal/ update e2e test naming

baldm0mma committed:
Merge branch 'main' into baldm0mma/transformation_info_modal

codeincarnate reviewed (approved):
Took another look and looks great 🥳. I think we should probably move from a modal to a drawer since we're doing that in general with the transformation interfaces, but we can leave that for a follow-up PR.
